framework/include/utils/IndexableProperty.h

```diff
@@ -0,0 +1,125 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "MooseTypes.h"
+#include <vector>
+
+template <typename T, bool is_ad>
+class IndexableProperty
+{
+public:
+  static InputParameters validParams();
+
+  IndexableProperty(T * host,
+                    const std::string & mat_prop_param = "mat_prop",
+                    const std::string & component_param = "component");
+
+  GenericReal<is_ad> operator[](int i) const;
+  operator bool() const;
+
+protected:
+  template <typename P>
+  void getPropertyHelper(const GenericMaterialProperty<P, is_ad> *& pointer,
+                         unsigned int components);
+
+  const GenericMaterialProperty<Real, is_ad> * _mat_prop_real;
+  const GenericMaterialProperty<RealVectorValue, is_ad> * _mat_prop_real_vector_value;
+  const GenericMaterialProperty<RankTwoTensor, is_ad> * _mat_prop_rank_two_tensor;
+  const GenericMaterialProperty<RankThreeTensor, is_ad> * _mat_prop_rank_three_tensor;
+  const GenericMaterialProperty<RankFourTensor, is_ad> * _mat_prop_rank_four_tensor;
+
+  T * _host;
+
+  const std::string & _mat_prop_param;
+  const std::string & _mat_prop_name;
+  const std::string & _component_param;
+
+  const std::vector<unsigned int> _component;
+};
+
+template <typename T, bool is_ad>
+InputParameters
+IndexableProperty<T, is_ad>::validParams()
+{
+  auto params = T::validParams();
+  params.template addRequiredParam<MaterialPropertyName>("mat_prop",
+                                                         "The name of the material property");
+  params.template addParam<std::vector<unsigned int>>(
+      "component",
+      "Index vector of the scalar component to extract from "
+      "the material property (empty for scalar properties)");
+  return params;
+}
+
+template <typename T, bool is_ad>
+IndexableProperty<T, is_ad>::IndexableProperty(T * host,
+                                               const std::string & mat_prop_param,
+                                               const std::string & component_param)
+  : _host(host),
+    _mat_prop_param(mat_prop_param),
+    _mat_prop_name(_host->template getParam<MaterialPropertyName>(_mat_prop_param)),
+    _component_param(component_param),
+    _component(host->template getParam<std::vector<unsigned int>>(_component_param))
+{
+  getPropertyHelper<Real>(_mat_prop_real, 0);
+  getPropertyHelper<RealVectorValue>(_mat_prop_real_vector_value, 1);
+  getPropertyHelper<RankTwoTensor>(_mat_prop_rank_two_tensor, 2);
+  getPropertyHelper<RankThreeTensor>(_mat_prop_rank_three_tensor, 3);
+  getPropertyHelper<RankFourTensor>(_mat_prop_rank_four_tensor, 4);
+}
```

Comment:
My first thought when I saw this is that we need to check to make sure we don't have more than one of these properties, but I guess MOOSE won't allow us to have properties of different types with the same name, right?

Reply:
Correct!